### PR TITLE
Calculate average values for rubric questions #4603

### LINKED ISSUE #4603: Calculate average values for rubric questions
Ideally, users should be able to attach different numeric values to each step of the rubric. But in the meantime, may be we can imply numeric values and calculate averages?

### PR DESCRIPTION
**Pull request by Sumei1009**
Fixes #4603

### CODE REVIEW

**Comment by LowWeiLin** [comment]
Can you provide some screenshots and explanations of what the changes do?

**Comment by Sumei1009** [comment]
If the user does not assign the weight to rubric choices: 
screen shot 2016-02-04 at 14 37 55
screen shot 2016-02-04 at 14 39 03

**Comment by Sumei1009** [comment]
If the user assign value: 
screen shot 2016-02-04 at 14 40 01
screen shot 2016-02-04 at 14 40 19
screen shot 2016-02-04 at 14 41 38

**Comment by Sumei1009** [comment]
I am still trying to figure a better way to allow user to assign value. Can I create a new issue for it?

**Comment by LowWeiLin** [comment]
I think this way is fine. What other methods do you suggest?
What happens when not all options are assigned values?
Also, is there a reason why the default values are in descending order?

**Comment by Sumei1009** [comment]
Maybe we can add a button that can select the assigned value to each choice? I am not sure whether it will be convenient for the user. 
If not all options are assigned value, the default value will be assigned, and those choices assigned value will be treated as not-assigned value but a normal choice. 
The reason why in descending order is because the default choice are strongly agree, agree, disagree and strongly disagree which the degree of agreement is descending.

**Comment by damithc** [comment]
1. Assigning values by user can be done as a separate issue. For this issue, assign values by default.
2. Change default values to be ascending, and change the order of default labels to go from `strongly disagree` to `strongly agree`

**Comment by thyageshm** [comment]
@Sumei1009 could we try and wrap this up today?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 1]
```diff
@@ -214,15 +214,27 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
...
+        }else{
+            for(int i = 0 ; i < numOfRubricChoices ; i++) {
+                String tableHeaderCell = 
+                        FeedbackQuestionFormTemplates.populateTemplate(tableHeaderFragmentTemplate,
+                                "${qnIndex}", questionNumberString,
+                                "${respIndex}", responseNumberString,
+                                "${col}", Integer.toString(i),
+                                "${rubricChoiceValue}", rubricChoices.get(i));
+                tableHeaderFragmentHtml.append(tableHeaderCell + Const.EOL);
+                }
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: The coding style is off in a few aspects here:
- `if (condition) {` and `for (loop) {` notice the space before/after the brackets
- `} else {` ditto
- `(a - b)` ditto, before/after the minus
- `int i = 0; i < ...; i++` no whitespace before the semicolon
- The closing `}` should be on the same indentation level as the starting `for` keyword.

File: src/main/resources/feedbackQuestionRubricResultStatsTemplate.html [hunk 2]
```diff
@@ -13,6 +13,7 @@
                     <tr>
                         <th></th>
                         ${tableHeaderRowFragmentHtml}
+                        <th><p>Average</p></th>
```
Comment (wkurniawan07) [inline src/main/resources/feedbackQuestionRubricResultStatsTemplate.html]: I don't think you need `<p>...</p>` wrapper inside a table cell.

File: src/main/resources/feedbackQuestionRubricResultStatsAverageFragment.html [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<td>
+    ${average}
+</td>
```
Comment (wkurniawan07) [inline src/main/resources/feedbackQuestionRubricResultStatsAverageFragment.html]: Can this be moved somewhere? It looks really really odd to have a template file with just 3 lines.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 4]
```diff
@@ -214,15 +214,27 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
...
-            tableHeaderFragmentHtml.append(tableHeaderCell + Const.EOL);
+        
+        if(getChoiceAssginedWeight(rubricChoices) == null){
+            for(int i = 0 ; i < numOfRubricChoices ; i++) {
+                String tableHeaderCell = 
+                        FeedbackQuestionFormTemplates.populateTemplate(tableHeaderFragmentTemplate,
+                                "${qnIndex}", questionNumberString,
+                                "${respIndex}", responseNumberString,
+                                "${col}", Integer.toString(i),
+                                "${rubricChoiceValue}", (rubricChoices.get(i) + "(" + (numOfRubricChoices-i) + ")"));
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Add a space before the opening bracket for better readability (not code-wise, but usability wise, i.e `Choice (1)` should be better than `Choice(1)`).

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 5]
```diff
@@ -728,5 +791,24 @@ public String getQuestionTypeChoiceOption() {
         return errors;
     }
 
+    private int[] getChoiceAssginedWeight(List<String> rubricChoices){
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: This method could be better named as `getWeightsAssignedToChoices`.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 6]
```diff
@@ -728,5 +791,24 @@ public String getQuestionTypeChoiceOption() {
...
+        int[] assignedWeight = new int[rubricChoices.size()];
+        for(int i=0; i<rubricChoices.size(); i++){
+            int length = rubricChoices.get(i).length();
+            if(rubricChoices.get(i).indexOf(")") == -1 ||rubricChoices.get(i).indexOf(")") != length-1){
+                return null;
+            }
+            try{
+                //assume the assigned number is between 0-99
+                //TODO: Figure out a better way to allow user to assign number by themselves
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: This TODO has no direct relation to what this method is concerned.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 7]
```diff
@@ -728,5 +791,24 @@ public String getQuestionTypeChoiceOption() {
...
+                int indexOfOpenParen = rubricChoices.get(i).indexOf('(', length - 4);
+                int indexOfCloseParen = rubricChoices.get(i).indexOf(')', length - 4);
+                assignedWeight[i] = Integer.parseInt(rubricChoices.get(i).substring(indexOfOpenParen+1, indexOfCloseParen));
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: This looks very hackish and dangerous. What's is trying to do?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 8]
```diff
@@ -542,6 +572,35 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+            int[] rubricChoiceWeight = getChoiceAssginedWeight(fqd.rubricChoices);
+            if(rubricChoiceWeight != null){
+                for(int j=0; j<fqd.numOfRubricChoices ; j++){
+                    average[i] += rubricChoiceWeight[j] * percentageFrequency[i][j];
+                }
+            }else{
+                for(int j=0; j<fqd.numOfRubricChoices ; j++){
+                    average[i] += (fqd.numOfRubricChoices - j) * percentageFrequency[i][j];
+                }
+            }
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: You can collapse this block to just:
```java
average[i] += (rubricChoiceWeight == null ? ... : ...) * ...;
```

**Comment by wkurniawan07** [comment]
We still have a not-so-long-but-not-so-short way to go, but good job on taking this issue :+1: The improvements suggested by @damithc can be done separately.
@thyageshm this can only be done by earliest next milestone. On @Sumei1009's defense, no one was reviewing until I stepped in just now.

**Comment by thyageshm** [comment]
Ok! we can keep this for next release :-)

File: src/main/resources/feedbackQuestionRubricResultStatsTemplate.html (hunk 2 above)
Comment (Sumei1009) [inline src/main/resources/feedbackQuestionRubricResultStatsTemplate.html]: If <p>..</p> is not there, the position of the header is inconsistent.

**Comment by Sumei1009** [comment]
I update some code according to the review. I will continue working on allowing user to assign value.

**Comment by wkurniawan07** [comment]
Sorry for making you undoing your work, but changing the order of the choices shouldn't be part of this issue... if you do more stuff than what is needed from a PR, it makes it harder for the reviewer (in this case, me) to review the changes.

**Comment by wkurniawan07** [comment]
Also, it seems like you're doing coding style fixes in parts not related to your work. As much as fixing coding style is appreciated, it'll make the PR harder to review and also harder to handle when a "revert" is needed in the future.

**Comment by Sumei1009** [comment]
Then I should work on another issue of changing the order? I will revert the code back first.

**Comment by wkurniawan07** [comment]
Yup, that will be the right thing to do (:

**Pull request closed without merging**